app/javascript/src/AccessTokens/components/ExpirationDatePicker.tsx

```diff
@@ -0,0 +1,126 @@
+
+import { createReactWrapper } from 'utilities/createReactWrapper'
+import type { IRecord } from 'utilities/patternfly-utils'
+
+import type { FunctionComponent, FormEvent } from 'react'
+
+import './ExpirationDatePicker.scss'
+
+interface ExpirationItem extends IRecord {
+  id: number;
+  name: string;
+  period: number; // In seconds
+}
+
+const collection: ExpirationItem[] = [
+  { id: 1, name: '7 days', period: 7 },
+  { id: 2, name: '30 days', period: 30 },
+  { id: 3, name: '60 days', period: 60 },
+  { id: 4, name: '90 days', period: 90 },
+  { id: 5, name: 'Custom...', period: 0 },
+  { id: 6, name: 'No expiration', period: 0 }
+]
+
+const dayMs = 60 * 60 * 24 * 1000
+const msExp = /\.\d{3}Z$/
+
+interface Props {
+  id: string;
+  label: string | null;
+}
+
+const ExpirationDatePicker: FunctionComponent<Props> = ({ id, label }) => {
+  const [selectedItem, setSelectedItem] = useState(collection[0])
+  const [pickedDate, setPickedDate] = useState(new Date())
+  const fieldName = `human_${id}`
+  const fieldLabel = label ?? 'Expires in'
+
+  const fieldDate = useMemo(() => {
+    if (selectedItem.period === 0) return null
+
+    return new Date(new Date().getTime() + selectedItem.period * dayMs)
+  }, [selectedItem])
+
+  const fieldHint = useMemo(() => {
+    if (!fieldDate) return
+
+    const date = new Date(fieldDate)
+    date.setHours(0, 0, 0, 0)
+
+    return `The token will expire on ${date.toLocaleDateString()}`
+  }, [fieldDate])
+
+  const dateValue = useMemo(() => {
+    let value = ''
+
+    if (fieldDate) {
+      value = fieldDate.toISOString().replace(msExp, 'Z')
+    } else if (selectedItem.id === 5 ) {
+      value = pickedDate.toISOString().replace(msExp, 'Z')
+    }
+
+    return value
+  }, [fieldDate, selectedItem, pickedDate])
+
+  const handleOnChange = (_value: string, event: FormEvent<HTMLSelectElement>) => {
+    const value = (event.target as HTMLSelectElement).value
+    const selected = collection.find(i => i.id.toString() === value) ?? null
+
+    if (selected === null) return
+
+    setSelectedItem(selected)
+    setPickedDate(new Date())
+  }
+
+  return (
+    <>
+      <FormGroup
+        isRequired
+        fieldId={fieldName}
+        label={fieldLabel}
+      >
+        <FormSelect
+          className="pf-c-form-control-expiration"
+          id={fieldName}
+          value={selectedItem.id}
+          onChange={handleOnChange}
+        >
+          {collection.map((item: ExpirationItem) => {
+            return (
+              <FormSelectOption
+                key={item.id}
+                label={item.name}
+                value={item.id}
+              />
+            )
+          })}
+        </FormSelect>
+        <span className="pf-c-form-control-expiration-hint">{fieldHint}</span>
+      </FormGroup>
+      <input id={id} name={id} type="hidden" value={dateValue} />
+      {selectedItem.id === 5 && (
```

Comment:
I don't love these synthetic ids on the options list. It's a bit hard to understand why this is checked for equality with `5`.

Some alternative ideas I had: 
- use a string instead of the numeric ID for an easier interpretation, e.g.
```
  { id: '7', name: '7 days', period: 7 },
  { id: '30', name: '30 days', period: 30 },
  { id: 'custom', name: 'Custom...', period: 0 },
  { id: 'no-exp', name: 'No expiration', period: 0 }
```

or something like that.

Or even remove the `id`, and reuse `period`, and have for example `period: 0` for "no expiration", and `period: -1` for custom. This way there are less values that we need to care about.

I also suggest to use `label` rather than `name`, as it is used for the field `label` in `FormSelectOption`, so we'll have something like
```
 <FormSelectOption
    key={item.period}
    label={item.label}
    value={item.period}
  />

Reply:
> use a string instead of the numeric ID for an easier interpretation, e.g.

Done https://github.com/3scale/porta/pull/3943/commits/fda4ff27cc8055c67523056c3ce401910da82670

> Or even remove the id, and reuse period, and have for example period: 0 for "no expiration", and period: -1 for custom. This way there are less values that we need to care about.

I didn't do this because I find this as confusing as the previous code.

> I also suggest to use label rather than name, as it is used for the field label in FormSelectOption, so we'll have something like

Done https://github.com/3scale/porta/pull/3943/commits/14b083d8d4eae1262f6f043a37ab0e349acd4fe3